### PR TITLE
C# generator fixes for array misinterpretation and size miscalculation

### PR DESCRIPTION
**Pull request by rickpresley**
This fixes two errors in generated C# code:

- Arrays of certain data types were being misinterpreted as other data types.
- Generated deserialization methods weren't properly calculating the size of arrays, which caused wrong offsets to be calculated for subsequent data fields

### CODE REVIEW

**Comment by mday299** [comment]
+1

**Comment by mday299** [comment]
@LorenzMeier *bump*

File: pymavlink/generator/mavgen_cs.py [hunk 1]
```diff
@@ -138,7 +138,7 @@ def generate_Deserialization(outf, messages):
         for f in m.ordered_fields:
             if (f.array_length):
                 outf.write("\t\t\t\t%s =  ByteArrayUtil.%s(bytes, offset + %s, %s),\n" % (mapFieldName.get(f.name, f.name), mapType[f.type][0], offset, f.array_length))
-                offset += f.array_length
+                offset += (f.array_length * mapType[f.type][1])
```
Comment (magicrub) [inline pymavlink/generator/mavgen_cs.py]: all the other changes are obviously good, this one I'm not sure what it's doing. The offset possibly changes, was it broken before? What is [1]?

Comment (rickpresley) [inline pymavlink/generator/mavgen_cs.py]: f.type indicates the type of data in the array.
mapType[f.Type] gets a record that holds relevant information for the data type.
mapType[f.Type][1] gives the size in bytes of the data type.

See mavlink/pymavlink/generator/mavgen_cs.py, line 29, for the definition of mapType.

Before, the code was taking the element count of the array and treating that as the size in bytes of the array.  That's fine if your array has byte-size data types in it, but if you have things like floats or doubles, the calculated array size is much too small and the offsets of subsequent fields wind up wrong.

Comment (magicrub) [inline pymavlink/generator/mavgen_cs.py]: yeah, that's what I was figuring it was. Just wanted to make sure .array_length didn't always assume byte-sized array. I didn't look at the whole file to investigate the structure definition.

LGTM

**Comment by mday299** [comment]
@magicrub  - Thanks for reviewing the PR. Do you know if we reach out to anyone in addition to @LorenzMeier for possibly getting this merged into master?

**Comment by magicrub** [comment]
I'll ping Tridge on it.

**Comment by mday299** [comment]
@magicrub -- while you're at it, can you have @tridge look at https://github.com/mavlink/mavlink/pull/584 as well?

Thanks,
-m

File: pymavlink/generator/mavgen_cs.py (hunk 1 above)
Comment (rickpresley) [inline pymavlink/generator/mavgen_cs.py]: Thanks much!

**Comment by magicrub** [comment]
currently has a merge conflict.

**Comment by tridge** [comment]
applied, thanks!

**Pull request closed without merging**